STM32/FLASH_readwrite/Src/FLASH_readwrite.c

```diff
@@ -1,140 +1,248 @@
  *  Description:		Provides an interface to read and write FLASH memory.
- *  TODO: This is a very brute force implementation where no 'protocol' is used
- *        to read/write the flash. So flash layout should be defined.
 */
-#if defined(STM32F401xC)
 
-#include <stm32f4xx_hal.h>
+#if defined(STM32F401xC)
 #include <FLASH_readwrite.h>
 #include <string.h>
 
-// NOTE: The flash sector and addr variables originate from the ld linker script.
-// The flash is used for program data and this must NOT be cleared. Putting data in
-// linker script will ensure that program data will not be cleared when writing
-// calibration data to the flash. To access these variables the syntax below is used.
-extern uint32_t _FlashSector; // Variable defined in ld linker script.
-extern uint32_t _FlashAddr;   // Variable defined in ld linker script.
-#define FLASH_SECTOR ((uint32_t) &_FlashSector)
-#define FLASH_ADDR   ((uint32_t) &_FlashAddr)
-
 #ifdef __cplusplus
 extern "C" {
 #endif
 
-// Write data to FLASH_ADDR+indx directly. This method does not
-// ensure any data integrity as data validation is performed.
-void writeToFlash(uint32_t indx, uint32_t size, uint8_t *data)
+/***************************************************************************************************
+** PRIVATE FUNCTION PROTOTYPES
+***************************************************************************************************/
+
+static uint32_t getFlashSector(uint32_t address);
+static int isWriteWithinSector(uint32_t address, uint32_t size);
+static int isProgramMemory(uint32_t address, uint32_t size);
+
+/***************************************************************************************************
+** PRIVATE FUNCTIONS
+***************************************************************************************************/
+
+/*!
+** @brief Gets the FLASH sector of an address.
+** @return Returns the FLASH sector if valid and 10 if a non-valid address is supplied.
+*/
+static uint32_t getFlashSector(uint32_t address)
+{
+    uint32_t address_sectors_start[VALID_SECTORS] = {START_SECTOR_1, START_SECTOR_2, START_SECTOR_3,
+                                                     START_SECTOR_4, START_SECTOR_5};
+
+    uint32_t address_sectors_end[VALID_SECTORS] = {END_SECTOR_1, END_SECTOR_2, END_SECTOR_3,
+                                                   END_SECTOR_4, END_SECTOR_5};                                         
+    
+    for (int i = 0; i < VALID_SECTORS; i++)
+    {
+        if (address >= address_sectors_start[i] &&
+            address <= address_sectors_end[i])
+        {
+            return i+1;
+        }
+    }
+    // This should never occur - return non-existing sector.
+    return 10;
+}
+
+/*!
+** @brief Checks that the data to be stored is contained within a single FLASH sector.
+** @return Returns 0 if the data is contained within a single FLASH sector, otherwise 1.
+*/
+static int isWriteWithinSector(uint32_t address, uint32_t size)
```

Comment:
From the name of the function, I would guess that it returns 1 if the data is contained in one sector but it does the opposite. Also why not using a boolean as the int will be either 0 or 1?

Reply:
We would have to `#include <stdbool.h>` and personally I prefer int over bool, but that's just a personal preference